addons/account/models/account_account.py

```diff
@@ -858,6 +980,272 @@ def get_import_templates(self):
     def _merge_method(self, destination, source):
         raise UserError(_("You cannot merge accounts."))
 
+    def action_merge(self):
+        """ Merge the accounts in `self`:
+            - the first one is extended to each company of the accounts in `self`, keeping their codes and names;
+            - the others are deleted; and
+            - journal items and other references are retargeted to the first account.
+
+            If the companies of several accounts overlap, this is an accounting operation, so we check that no impacted entries are in a locked period.
+            If the companies don't overlap, from an accounting perspective, each company still has its own independent view of the account.
+        """
+        # Step 1: Checks.
+        if len(self) < 2:
+            raise UserError(_("You must select at least 2 accounts to merge."))
+
+        for field in ['currency_id', 'deprecated', 'account_type', 'reconcile', 'non_trade']:
+            if len(set(self.mapped(field))) > 1:
+                raise UserError(_(
+                    "You may only merge accounts that have the same account type, currency, deprecated status, "
+                    "reconciliation status, and trade/non-trade receivable status."
+                ))
+
+        # Check that if any companies overlap, no impacted entries are in a locked period.
+        accounts_by_company = defaultdict(lambda: self.env['account.account'])
+        for account in self:
+            for company in account.company_ids:
+                accounts_by_company[company] |= account
+
+        # If exactly one of the accounts contains locked entries, we will merge the other ones into it.
+        account_to_merge_into = self.env['account.account']
+
+        is_irreversible = False
+        bad_entries = {}
+        for company, accounts in accounts_by_company.items():
+            if len(accounts) > 1:
+                is_irreversible = True
+
+                domain = [
+                    ('account_id', 'in', (accounts - account_to_merge_into).ids),
+                    ('parent_state', '=', 'posted'),
+                ]
+                bad_entries_current_company = {}
+                if hashed_items := self.env['account.move.line'].sudo().search([*domain, ('move_id.inalterable_hash', '!=', False)]):
```

Comment:
I don't think we need to know exactly which entry is causing an issue; we just need to know if at least one will be. (otherwise, perf issue)
There is nothing the user can do about it anyway...